### ISSUE TITLE
Page Break Issue

### ISSUE THREAD

**Issue by DarshanKsheerasagar** [wkhtmltopdf/wkhtmltopdf#3422]
Hello All ,

Generate PDF using HTML string with page break is not proper working ... Below is my Code and Output

Default.aspx

<%@ Page Language="C#" AutoEventWireup="true" CodeBehind="Default.aspx.cs" Inherits="wkhtmltopdfTest._Default" %>

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head runat="server">
    <title></title>
    <link href="PrintStyle.css" rel="stylesheet" type="text/css" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
</head>
<body>
    <form id="form1" runat="server">
     <asp:Button ID="Button1" runat="server" Text="Download" OnClick="btnExport_Click" />
    <div >
        <asp:Panel ID="pnlPerson" runat="server">
            <div>
                <table style="width: 785px;">
                    <tr>
                        <td>
                            First Page
                        </td>
                    </tr>
                    <tr class="page-break">
                        <td >
                        </td>
                    </tr>
                    <tr>
                        <td>
                            Second Page
                        </td>
                    </tr>
                    <tr class="page-break">
                        <td >
                        </td>
                    </tr>
                    <tr>
                        <td>
                            Third Page
                        </td>
                    </tr>
                    <tr class="page-break">
                        <td >
                        </td>
                    </tr>
                    <tr>
                        <td>
                            Fourth Page
                        </td>
                    </tr>
                    <tr class="page-break">
                        <td >
                        </td>
                    </tr>
                    <tr>
                        <td>
                            Five Page
                        </td>
                    </tr>
                </table>
            </div>
        </asp:Panel>
    </div>
   
    </form>
</body>
</html>

Default.cs.aspx

using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace wkhtmltopdfTest
{
    public partial class _Default : System.Web.UI.Page
    {
        protected void Page_Load(object sender, EventArgs e)
        {
        }
        protected void btnExport_Click(object sender, EventArgs e)
        {
            try
            {
                string PrintStyle = System.IO.File.ReadAllText(Server.MapPath("/PrintStyle.css"));

                StringWriter sw = new StringWriter();
                HtmlTextWriter hw = new HtmlTextWriter(sw);

                pnlPerson.RenderControl(hw);
                StringReader sr = new StringReader(sw.ToString());

                string outXml = "<html> <head> " +
                                "<style type=\"text/css\">  " + PrintStyle + " </style> " +
                                "</head> <body> <div>" + sw.ToString() + " " +
                                " </div></body> </html>";
                //table, tr, td, th, tbody, thead, tfoot { page-break-inside: avoid !important;}  
                //tr{display:block;}

                byte[] result = ConvertHtmlToPdf(outXml);

                HttpContext.Current.Response.Clear();
                Response.ContentType = "application/pdf";
                string outputFilename = "Report" + "_" + DateTime.Now.ToString("yyyy-MM-dd-hh-mm-ss-fff") + ".pdf";
                Response.AddHeader("content-disposition", "attachment;filename=" + outputFilename);
                Response.CacheControl = "No-cache";
                MemoryStream ms = new MemoryStream(result);
                ms.WriteTo(Response.OutputStream);
                Response.Flush();
                Response.SuppressContent = true;
                HttpContext.Current.ApplicationInstance.CompleteRequest();

            }
            catch(Exception ex)
            {
            }
        }
        protected byte[] ConvertHtmlToPdf(string html)
        {
            System.Reflection.Assembly ass = System.Reflection.Assembly.GetExecutingAssembly();
            string conPath = System.IO.Path.GetDirectoryName(ass.CodeBase);
            Uri uri = new Uri(conPath);
            conPath = uri.AbsolutePath.Replace("%20", " ").Replace("25", "");
            conPath = Path.GetFullPath(Path.Combine(conPath, @"..\"));
            conPath = conPath + "bin\\wkhtmltopdf.exe";
            var p = new Process
            {
                StartInfo =
                {
                    CreateNoWindow = true,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    RedirectStandardInput = true,
                    UseShellExecute = false,
                    FileName = conPath
                }
            };
            string s = "-q ";
            s += "--margin-bottom 0 ";
            s += "--margin-right 0 ";
            s += "--margin-left 0 ";
            s += "--margin-top 0 ";
            s += "--page-size A4 ";
            s += "--print-media-type ";
            s += "--disable-smart-shrinking ";
            s += "- -";
            p.StartInfo.Arguments = s;
            p.Start();
            StreamWriter myStreamWriter = new StreamWriter(p.StandardInput.BaseStream, Encoding.UTF8);
            myStreamWriter.Write(html);
            myStreamWriter.Close();
            var buffer = new byte[32768];
            byte[] file;
            using (var ms = new MemoryStream())
            {
                while (true)
                {
                    var read = p.StandardOutput.BaseStream.Read(buffer, 0, buffer.Length);
                    if (read <= 0)
                    {
                        break;
                    }
                    ms.Write(buffer, 0, read);
                }
                file = ms.ToArray();
            }
            p.WaitForExit(60000);
            var returnCode = p.ExitCode;
            p.Close();
            return returnCode == 0 ? file : null;
        }
    }
}

PrintStyle.css

 @media print 
 { 
     tr.page-break  {display: block;page-break-after: always; }
 }

Output :

![pagebreakissue](https://cloud.githubusercontent.com/assets/24871149/24740381/7a48f3e6-1abe-11e7-91c9-ed947c8ae5d6.JPG)

**Comment by PhilterPaper** [comment]
Darshan, I can't understand any of your posts. Please learn to write in plain English, and not extremely cryptic and terse Twitter 140 character mode. Twits are for morons.

Evidently you are have some sort of problem with pages, but you don't say exactly what the trouble is. I can't magnify the image enough to see what you've circled. I'm going to take a wild guess that what should be on the first line of each page is appearing further and further down the page, but without seeing the HTML in question, I can't tell any more than that.

**Comment by DarshanKsheerasagar** [comment]
Hello @PhilterPaper ,

I am Generating PDF from input is HTML string and it generating the pdf also but the problem is PAGE BREAK. i have applied  "tr.page-break {display: block;page-break-after: always; }" property like this its doing page break but some while line block white space coming in the page so its not PROPER PAGE BREAK Coming . 

I Have attached PDF File You can see and see the page preak not proper.


[Report_2017-04-07-11-06-46-009.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/904614/Report_2017-04-07-11-06-46-009.pdf)

**Comment by PhilterPaper** [comment]
For the many among us who don't speak C#, what is the HTML being converted, as well as any JS and CSS? It sounds like you have a `<tr class="page-break">`, but that's not enough information. Have you tried using divs instead of table layout?

**Comment by anandhuded2** [comment]
Hi Darhsan,

If you found any solution to it please mention it over here because I am also facing same problem.
